src/context.js

```diff
@@ -0,0 +1,191 @@
+    return this._promise;
+  }
+
+  finalizePromise() {
+    if(this.isCompleted && this.resolve) {
+      this.resolve(this.result);
+    } else if(this.isErrored && this.reject) {
+      this.reject(this.result);
+    } else if(this.isHalted && this.reject) {
+      this.reject(new HaltError(this.result));
+    }
+  }
+
+  then(...args) {
+    return this.promise.then(...args);
+  }
+
+  catch(...args) {
+    return this.promise.catch(...args);
+  }
+
+  finally(...args) {
+    return this.promise.finally(...args);
+  }
+
+  get root() {
+    if (!this.parent) {
+      return this;
+    } else {
+      return this.parent.root;
+    }
+  }
+
+  call(operation, options = {}) {
+    let child = new ExecutionContext(this, options);
+    this.children.add(child);
+    child.ensure(() => {
+      this.children.delete(child);
+      let { result } = child;
+      if (this.isRunning || this.isWaiting) {
+        if (child.isErrored) {
+          if (typeof options.fail === 'function') {
+            options.fail(result, this.fail);
+          } else {
+            this.fail(result);
+          }
+        } else if (child.isHalted) {
+          if (typeof options.halt === 'function') {
+            options.halt(result);
+          } else {
+            this.fail(new HaltError(result));
+          }
+        }
+        if (this.isWaiting && !this.hasBlockingChildren) {
+          this.finalize('completed');
+        }
+      }
+    });
+    child.enter(operation);
+    return child;
+  }
+
+  ensure(hook) {
+    let { exitHooks } = this;
+    exitHooks.add(hook);
+    return () => exitHooks.delete(hook);
+  }
+
+  enter(operation) {
+    if (this.isUnstarted) {
+      let controller = this.createController(operation);
+      this.operation = operation;
+      this.state = 'running';
+
+      let { resume, fail, ensure, call } = this;
+      controller.call({ resume, fail, ensure, call, context: this });
+
+    } else {
+      throw new Error(`
+Tried to call #enter() on a Context that has already been finalized. This
+should never happen and so is almost assuredly a bug in effection. All of
+its users would be in your eternal debt were you to please take the time to
+report this issue here:
+https://github.com/thefrontside/effection.js/issues/new
+
+Thanks!`);
+    }
+  }
+
+  halt(reason) {
+    if (this.isBlocking) {
+      this.finalize('halted', reason);
+    }
+  }
+
+  resume(value) {
+    if (this.isRunning) {
+      this.result = value;
+      if (this.hasBlockingChildren) {
+        this.state = 'waiting';
+      } else {
+        this.finalize('completed', value);
+      }
+    } else {
+      throw new Error(`
+Tried to call Fork#resume() on a Fork with state '${this.state}' This
+should never happen and so is almost assuredly a bug in effection. All of
+its users would be in your eternal debt were you to please take the time to
+report this issue here:
+https://github.com/thefrontside/effection.js/issues/new
+
+Thanks!`);
```

Comment:
Is it just me or is this not actually true anymore? `resume` is part of the public API now, so someone creating a controller could very well call `resume` twice, which would still be wrong, but not a bug in effection?

Reply:
Since this is similar to the Promise API, it might be worth mimicking how that API works, where I think resolving after the first time is simply a no-op.

Reply:
Nice catch, not true at all.

Reply:
Should it be a warning?

Reply:
With promises, this is a no-op without warning, so I think that behaviour is acceptable.